File: MoppLib/MoppLib/PublicInterface/MoppLibCardReaderManager.m
//
//  MoppLibCardReaderManager.m
//  MoppLib
//
/*
 * Copyright 2017 - 2024 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#import "MoppLibCardReaderManager.h"
#import "CardActionsManager.h"
#import "CardReaderiR301.h"
#import "ReaderInterface.h"

@interface MoppLibCardReaderManagerContext : NSObject
@property (nonatomic) SCARDCONTEXT handle;
@end

@implementation MoppLibCardReaderManagerContext

-(instancetype)init {
    if (self = [super init]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &self->_handle);
            printLog(@"ID-CARD: Started reader discovery: %x", self.handle);
        });
    }
    return self;
}

-(void)dealloc {
    SCARDCONTEXT copy = self.handle;
    dispatch_async(dispatch_get_main_queue(), ^{
        FtDidEnterBackground(1);
        SCardCancel(copy);
        SCardReleaseContext(copy);
        printLog(@"ID-CARD: Stopped reader discovery with status: %x", copy);
    });
}

@end

@interface MoppLibCardReaderManager()<ReaderInterfaceDelegate>
@property (nonatomic, strong) MoppLibCardReaderManagerContext *context;
@property (nonatomic, strong) ReaderInterface *readerInterface;
@property (nonatomic) MoppLibCardReaderStatus status;
@end

@implementation MoppLibCardReaderManager

+ (MoppLibCardReaderManager *)sharedInstance {
  static dispatch_once_t pred;
  static MoppLibCardReaderManager *sharedInstance = nil;
  dispatch_once(&pred, ^{
    sharedInstance = [[self alloc] init];
    sharedInstance.status = Initial;
    sharedInstance.readerInterface = [ReaderInterface new];
    [sharedInstance.readerInterface setDelegate:sharedInstance];
  });
  return sharedInstance;
}

- (void)startDiscoveringReaders {
    printLog(@"ID-CARD: Starting reader discovery");
    [self updateStatus:_status == Initial ? Initial : ReaderRestarted];
    self.context = [MoppLibCardReaderManagerContext new];
}

- (void)stopDiscoveringReaders {
    [self stopDiscoveringReadersWithStatus:Initial];
}

- (void)stopDiscoveringReadersWithStatus:(MoppLibCardReaderStatus)status {
    printLog(@"ID-CARD: Stopping reader discovery with status %lu", (unsigned long)status);
    CardActionsManager.sharedInstance.reader = nil;
    _status = status;
    self.context = nil;
}

- (void)updateStatus:(MoppLibCardReaderStatus)status {
    if (_status == status) {
        return;
    }

    _status = status;
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate)
            [self.delegate moppLibCardReaderStatusDidChange:status];
    });
}

#pragma mark - ReaderInterfaceDelegate

- (void)readerInterfaceDidChange:(BOOL)attached bluetoothID:(NSString *)bluetoothID {
    printLog(@"ID-CARD attached: %d", attached);
    if (!attached) {
        CardActionsManager.sharedInstance.reader = nil;
        return [self updateStatus:Initial];
    }
    CardActionsManager.sharedInstance.reader = [[CardReaderiR301 alloc] initWithContextHandle:self.context.handle];
    if (CardActionsManager.sharedInstance.reader != nil) {
        [self updateStatus:ReaderConnected];
    }
}

- (void)cardInterfaceDidDetach:(BOOL)attached {
    printLog(@"ID-CARD: Card (interface) attached: %d", attached);
    [self updateStatus:attached ? CardConnected : ReaderConnected];
}

- (void)didGetBattery:(NSInteger)battery {
}

- (void)findPeripheralReader:(NSString *)readerName {
    printLog(@"ID-CARD: Reader name: %@", readerName);
}

@end

File: MoppLib/MoppLib/PublicInterface/MoppLibCardReaderManager.swift
//
//  MoppLibCardReaderManager.swift
//  MoppLib
//
/*
 * Copyright 2017 - 2024 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

import iR301

public protocol MoppLibCardReaderManagerDelegate: AnyObject {
    func moppLibCardReaderStatusDidChange(_ status: MoppLibCardReaderStatus)
}

public enum MoppLibCardReaderStatus {
    case Initial
    case ReaderNotConnected
    case ReaderRestarted
    case ReaderConnected
    case CardConnected(CardCommands)
    case ReaderProcessFailed
}

public class MoppLibCardReaderManager {
    public static let shared = MoppLibCardReaderManager()

    public weak var delegate: MoppLibCardReaderManagerDelegate?
    fileprivate var handle: SCARDCONTEXT = 0
    private var handler = ReaderInterfaceHandler()
    private var status: MoppLibCardReaderStatus = .Initial

    private init() {
    }

    public func startDiscoveringReaders() {
        guard handle == 0 else {
            print("ID-CARD: Reader discovery is already running")
            return
        }
        print("ID-CARD: Starting reader discovery")
        updateStatus(status)
        SCardEstablishContext(DWORD(SCARD_SCOPE_SYSTEM), nil, nil, &handle)
        print("ID-CARD: Started reader discovery: \(handle)")
    }

    public func stopDiscoveringReaders(with status: MoppLibCardReaderStatus = .Initial) {
        print("ID-CARD: Stopping reader discovery")
        self.status = status
        FtDidEnterBackground(1)
        SCardCancel(handle)
        SCardReleaseContext(handle)
        print("ID-CARD: Stopped reader discovery with status: \(handle)")
        handle = 0
    }

    fileprivate func updateStatus(_ status: MoppLibCardReaderStatus) {
        self.status = status
        DispatchQueue.main.async {
            self.delegate?.moppLibCardReaderStatusDidChange(status)
        }
    }
}

private class ReaderInterfaceHandler: NSObject, ReaderInterfaceDelegate {
    private let readerInterface = ReaderInterface()

    override init() {
        super.init()
        readerInterface.setDelegate(self)
    }

    func readerInterfaceDidChange(_ attached: Bool, bluetoothID: String?) {
        print("ID-CARD attached: \(attached)")
        MoppLibCardReaderManager.shared.updateStatus(attached ? .ReaderConnected : .ReaderNotConnected)
    }

    func cardInterfaceDidDetach(_ attached: Bool) {
        print("ID-CARD: Card (interface) attached: \(attached)")
        guard attached, let reader = CardReaderiR301(contextHandle: MoppLibCardReaderManager.shared.handle) else {
            return MoppLibCardReaderManager.shared.updateStatus(.ReaderConnected)
        }
        do {
            let atr = try reader.powerOnCard()
            if let handler = Idemia(reader: reader, atr: atr) {
                MoppLibCardReaderManager.shared.updateStatus(.CardConnected(handler))
            }
        } catch {
            print("ID-CARD: Unable to power on card")
            MoppLibCardReaderManager.shared.updateStatus(.ReaderProcessFailed)
        }
    }

    func didGetBattery(_ battery: Int) {
        // Implement if needed
    }

    func findPeripheralReader(_ readerName: String) {
        print("ID-CARD: Reader name: \(readerName)")
    }
}
